Add config.d support

maybe-sybr opened the pull request:
This is an initial look at the work I've been doing related to #135.

maybe-sybr commented:
This doesn't have env var support yet for overriding stuff in the on-disk configs. Specifically, I want to be able to provide something like `MEDALLION_BACKEND_MONGO_URI` to override/provide the `.backend.uri` config key which the mongo backend uses. Trying to do that as a general system is annoying - although I've made some progress - but it would be much easier to simply teach the mongo backend itself about such an env-var so it can pull the var in itself. I'm planning to do that in my out-of-tree backend for the moment (using the changeset in #133 ) and then coming back at it when I have a bit more time to think it through.

maybe-sybr commented on setup.py:
```diff
@@ -48,8 +48,10 @@ def get_long_description():
     keywords="taxii taxii2 server json cti cyber threat intelligence",
     packages=find_packages(exclude=["*.test", "*.test.data"]),
     install_requires=[
+        "environ-config",
```
This changeset depends on a PR I have up for this package. If it doesn't get pulled then I have an alternative approach that isn't as nicely declarative but would work fine with the currently released versions of the environs package.

maybe-sybr replied:
This got merged into `environ-config` 21.1.0. Shall we set this `>=21.1`?

maybe-sybr commented:
The way I do environment variable config definitions is currently much easier with #133 merged in. I've done that here for the moment.

emmanvg commented:
It has been merged.

maybe-sybr commented:
Great, thanks! I'll work on cleaning this up and check in on if my PR in environ-config might be merged. If not, I'll pivot to an alternative approach to get this out of draft state and complete the outline I made in #135

maybe-sybr commented:
Apologies for the radio silence on this. It's still on my list and a little while ago my changes to `environ-config` were merged which should mean I can clean this up, rebase and de-draft it.

maybe-sybr commented:
I've now found the time to run through this and prep it for merge from a code perspective. The changeset is freshened on top of current master and I've fixed an issue with config being dropped when not specified in the environment. A couple of questions:
 * what's the expectation around testing? this is new code so I assume you might want some unit tests to cover it
 * @zrush-mitre suggested that this should be documented. I don't see anything config related in the `docs/` tree yet, will you need me to write something up?

maybe-sybr commented on docker-compose.yml:
```diff
@@ -7,7 +7,10 @@ services:
-    command: [sh, -c, "medallion --host 0.0.0.0 docker_config.json"]
+    command: [sh, -c, "medallion --host 0.0.0.0"]
+    env:
+      MEDALLION_BACKEND_MODULE_CLASS: "MongoBackend"
```
I do wish this could just be `MEDALLION_BACKEND: "MongoBackend"` but `environ-config` doesn't allow us to do variable name customisation like that :(

zrush-mitre reviewed (commented):
Hi @maybe-sybr,

Good job on the PR. Some interesting choices for programming, but functional nonetheless. I'll be doing some testing of it myself in the coming days.

For testing, we definitely want to maximize code coverage, but we don't have a coverage goal or anything. Rather we want to ensure that 1) common use cases are tested, 2) as much of the code is covered by tests (within reason) 3) edge and trouble cases are included and 4) the depth of the tests are increased, as compared to the tests that came before.

4 is really important, since often times when capabilities are added to code, they won't be heavily modified for some time, so we want to make sure that the types of tests that are included are encouraging other PRs to also increase their testing breadth/depth. The config tests before defined a few different uses and edge cases to make sure they had the correct behavior, but for this PR we want to see some config file variances, as well as testing the behaviors of the flowchart you made above to ensure things are and will work correctly.

Some documentation is definitely needed to show what the environment variables are, where the code looks for configs, and how they should be structured. Depending upon how in depth the documentation is, you can either make a new file in the docs folder, or add a straightforward section to the readme. Speaking of...

Make sure stuff like previous tests and previous documentation is modified as needed for your changes. The README may need some changes to show how to call initial servers and whatnot, and other tests may need the same changes.

Thanks for your work, and I'll continue reviewing your code.

maybe-sybr commented on medallion/config.py:
```diff
@@ -0,0 +1,82 @@
+            if conf_file is not DEFAULT_CONFFILE:
+                raise exc
+            else:
+                config_data = dict()
```
This can be undefined if we skip this when `conf_file is None`. I'm reworking it to be a bit more DRY and fix this bug.

maybe-sybr commented on medallion/config.py:
```diff
@@ -0,0 +1,82 @@
+    if conf_file is not None:
+        conf_file_p = pathlib.Path(conf_file)
+        try:
+            config_data = json.load(conf_file_p.open())
```
I also need to handle JSON decoding errors and raise something like a `TypeError` if the data we get back is not a JSON object (e.g. array, raw value) since json merge will complain about merging objects and other types and the config we get from `environ` is always going to be an object (ie. `dict`).

maybe-sybr reviewed (commented):
I'm most of the way through building out the tests for this. I should have something up before I finish working on it today.

maybe-sybr commented on medallion/config.py:
```diff
@@ -0,0 +1,82 @@
+    if conf_dir is not None:
+        conf_dir_p = pathlib.Path(conf_dir)
+        try:
+            for conf_file_p in conf_dir_p.iterdir():
```
This also needs to be sorted in some way since `iterdir()` won't do that for us

maybe-sybr commented:
Hmm, looks like maybe my trick for testing the argument parser doesn't work in 3.7 :/

zrush-mitre commented on medallion/test/test_startup.py:
```diff
@@ -0,0 +1,282 @@
+    parser = medallion.scripts.run._get_argparser()
+    parser.print_help()
+    (out, _) = capsys.readouterr()
+    assert "medallion v" in out
```
Was this supposed to be the whole text of the help command? I assume "medallion v" is just a random section of it, and that's okay, just seemed odd.

maybe-sybr replied:
It's just a reliable looking substring from the help text. The current version number follows the `v` at the end here and the test shouldn't care about that. It's mostly to validate that the custom formatter is doing its thing :)

zrush-mitre commented on medallion/test/config/test_config_loading.py:
```diff
@@ -0,0 +1,250 @@
+    Confirm an exception is raised if a non-default config file is missing.
+    """
+    conf_file = tmp_path / "missing.conf"
+    with pytest.raises(FileNotFoundError, match=repr(str(conf_file))):
```
Because of Windows paths, this needs an `re.escape` for the match, like so:

`match=re.escape(repr(str(conf_file))))`

maybe-sybr replied:
Ah good point! Did you run the tests on a Windows machine? I don't have any dev machines which I could use easily at the moment so if you are able to flush out any other issues which I would have missed because of that it would be helpful :) I'll amend this and the one you tagged below when I get some time tomorrow. Thanks

zrush-mitre replied:
Yeah, made sure to run this with both Unix and Windows. This seems like the only issue though (there may be some issues with environment variables, but I still have to test that).

maybe-sybr replied:
Made this change and just pushed an amended diff. Let me know if that works on your end. I tried to work out if it would play nicely using a variant of this test and passing in a `WindowsPurePath` but the new code creates a `Path` from whatever we pass in so there's some double escaping happening which makes it annoying for me to validate that the test might function as expected on Windows.

zrush-mitre commented on medallion/test/config/test_config_loading.py:
```diff
@@ -0,0 +1,250 @@
+    Confirm an exception is raised if a non-default config dir is missing.
+    """
+    conf_dir = tmp_path / "missing.d"
+    with pytest.raises(FileNotFoundError, match=repr(str(conf_dir))):
```
Because of Windows paths, this needs an `re.escape` for the match, like so:

`match=re.escape(repr(str(conf_file))))`

zrush-mitre commented on medallion/config.py:
```diff
@@ -0,0 +1,104 @@
+from . import backends
+
+DEFAULT_CONFFILE = "/etc/medallion.conf"
+DEFAULT_CONFDIR = "/etc/medallion.d/"
```
I'm not super familiar with configs and their standard places, but is there a better way we can make this platform agnostic? `/etc` is the obvious choice for Unix environments,  but this does not exist by default on Windows. Is there a file location that would work for both systems?

maybe-sybr replied:
Ah, good point. I should have twigged on that when you mentioned running the tests on Windows. IMO we'd normally want to use a base directory like `%APPDATA%/medallion/` or something similar. I might quiz a few Windows friends to see if they think that's insane or not.

maybe-sybr replied:
I've looked into this briefly and came across the appdirs library which seems like it'd be good to use, however it prefers to use XDG paths for Linux which means we'd get a default site config path of `/etc/xdg/medallion/` which would be disruptive for in-place upgrades. As an alternative, I could use the current paths for Linux and use appdirs on other platforms (Windows, macOS). Any thoughts?

zrush-mitre replied:
I'm personally against platform-dependent implementations, though I understand sometimes they are unavoidable. Any thoughts here @clenk?

maybe-sybr replied:
I've just pushed a commit which uses the `appdirs` library to provide sane paths for non-Linux platforms. That should give an idea of /a/ solution to the problem. Additionally, now or in future, we could add an opt-in to use XDG paths on Linux as well (ie. use appdirs on that platform too).

Happy for you to strip that commit off if you'd prefer not to use it, and I'll also keep an eye on this thread for input from you and @clenk.

maybe-sybr replied:
I meant that since the default path would go from `/etc/medallion.conf` in, e.g. the medallion docker image, to `/etc/xdg/medallion/<something>`. But you're right, I forgot that without this diff applied, medallion needs a specific path to a config file specified, and the docker image actually provides that path as `/opt/taxii/medallion.conf`.

I can certainly change this to use `appdirs` on all platforms and adjust any references to the `/etc/medallion{.conf,.d/}` paths.

maybe-sybr replied:
This is all done in the new diff, lmk what you think

zrush-mitre replied:
Looks really good actually. Would rather pass off the responsibility of configuration directories to that library than us handle it ourselves.

zrush-mitre commented on medallion/test/config/test_config_loading.py:
```diff
@@ -0,0 +1,253 @@
+    conf_file = tmp_path / "medallion.conf"
+    json.dump(confdata, conf_file.open("w"))
+    with subtests.test(msg="Used as config file"):
+        assert m_cfg.load_config(conf_file=conf_file) == confdata
```
Because this function checks environment variables, if variables are set when running the tests, it will cause a lot of them to fail. Is there a way to unset environment variables at test execution time, then reset them afterwards?

maybe-sybr replied:
I believe there is a pytest monkey_patch fixture which might do the job: https://docs.pytest.org/en/6.2.x/monkeypatch.html#monkeypatching-environment-variables

I'll look into some best practice for this - I think patching the `os.environ` dict might be reasonable but if that's not easy/recommended, we could potentially mock `environ_config` to pull from an alternative default source of variables.

zrush-mitre replied:
You could also get, delete, then reset them with a pytest fixture, though that's definitely more rough than using something like monkeypatch, if it works.

maybe-sybr replied:
Done, I sanity checked that the amended diff works with a poisonous envvar by applying the following patch. The previous diff would have all sorts of failures with the same envvar.

```diff
diff --git a/tox.ini b/tox.ini
index fb9be89..2d714ad 100644
--- a/tox.ini
+++ b/tox.ini
@@ -12,6 +12,8 @@ deps =
     responses
     pymongo
     pyjwt
+setenv =
+    MEDALLION_TAXII_MAX_PAGE_SIZE = 1337
 commands =
     pytest --cov=medallion medallion/test/ --cov-report term-missing
 
```

zrush-mitre commented on medallion/config.py:
```diff
@@ -0,0 +1,104 @@
+            config_data = jsonmerge.merge(config_data, new_data)
+    # Load extra configuration from the environment
+    env_config = MedallionConfig.from_environ().as_dict()
+    config_data = jsonmerge.merge(config_data, env_config)
```
While I do like the idea of combining the environment variables and the referenced config files, I can see a possibility of a user having issues when a variable is set or a file in the config folder that they don't know about, and they don't have a good grasp on this part of the config process. It may be worth it to print each config value and where it was retrieved from, either by default or when a CLI argument is set.

maybe-sybr replied:
That's a great idea, a `--conf-check` option or something could turn on some debugging in this function and then exit before actually running medallion?

maybe-sybr replied:
Done in the new commit at the tip of the branch

zrush-mitre commented on medallion/scripts/run.py:
```diff
@@ -57,11 +59,49 @@ def _get_argparser():
+    config_dir_group.add_argument(
+        "--no-conf-dir",
+        action="store_true",
+        help=inspect.cleandoc("""
```
Any reason the cleandoc was used here as opposed to the regular string from the other args?

maybe-sybr replied:
Huh, I don't even remember doing this. I probably just used it to get rid of the leading indentation and newlines so I could use long triple quoted strings without concatenating them myself

maybe-sybr commented on setup.py:
```diff
@@ -48,8 +48,11 @@ def get_long_description():
     keywords="taxii taxii2 server json cti cyber threat intelligence",
     packages=find_packages(exclude=["*.test", "*.test.data"]),
     install_requires=[
+        "appdirs>=1.4.4",
```
This is chosen because 1.4.4 was the first release they made which was marked as stable. It's still seeing slow development but no releases since mid last year.

maybe-sybr commented on medallion/config.py:
```diff
@@ -0,0 +1,162 @@
+    # On non-Linux systems we use and app specific config directory with an
+    # optional `config.d/` subdirectory
+    _appdirs = appdirs.AppDirs("medallion", "oasis-open")
+    DEFAULT_CONFFILE = pathlib.Path(_appdirs.site_config_dir)
```
This is a directory, not a file! Will fix.

maybe-sybr replied:
This is fixed in the new diff

maybe-sybr commented on medallion/config.py:
```diff
@@ -0,0 +1,173 @@
+# Find a config directory for the closest compatible version including the
+# current one, or fall back to the current version's one if none are extant
+_curr_version = packaging.version.Version(_medallion_version.__version__)
+_compat_versions = {_curr_version.major, 3, 2}
```
I included `_curr_version.major` as well as the explicit value `3` to ensure that v4 would try the v3 path unless some action is taken to avoid that (e.g. if a config schema change happens). I'm not sure what the most sensible dev flow for making a change like that would be, probably something like bumping the `medallion.version.__version__ = "4.0.0.dev0"`

zrush-mitre replied:
Would you rather do it this way than not including the `packaging` library and making `_compat_versions` equal to `{_medallion_version, "3", "2"}`? One less package and a little cleaner with the latter.

maybe-sybr replied:
Was that supposed to be `{_medallion_version.__version__, ...}`? That'd cause us to hunt for `/etc/xdg/medallion/X.Y.Z/` rather than just `.../X/` (being the current major version which will not be `3` in future).

The use of `packaging.version.Version` is to get canonical PEP440 version parsing rather than having to rely on something like `__version__.split(".", 1)[0]` which IMO is less semantically obvious.

Happy to go either way on this.

zrush-mitre replied:
I think you might've been right on the first try. Go with that for now and we can always change in the future if it becomes a dependency issue.

clenk commented:
Looks like this is why the build is failing, FYI:

> medallion/config.py:5:1: F401 'platform' imported but unused

maybe-sybr commented:
> Looks like this is why the build is failing, FYI:
> 
> > medallion/config.py:5:1: F401 'platform' imported but unused

Ah, remnant of the platform specific code which is gone now. I'll push an amended diff in a moment. Thanks!

zrush-mitre commented:
Hey @maybe-sybr,

Don't wanna leave you in the dark, so here's a quick update:

@clenk is off for some time, so I'm getting promoted to maintainer. Once I have permissions here, I'm gonna give another last review before merging this in. We appreciate all the effort you put in here!

Thanks!

maybe-sybr commented:
Thanks for the update, @zrush-mitre. Much appreciated!

zrush-mitre commented on README.rst:
```diff
@@ -76,17 +76,35 @@ Medallion provides a command-line interface to start the TAXII Server
+                            value of the MEDALLION_CONFFILE environment variable
+                            or /etc/xdg/medallion/3/medallion.conf.
+
+      --conf-dir CONF_DIR   Path a directory containing JSON configuration files
```
"Path a directory" should be "Path to a directory"

maybe-sybr replied:
fixed here and also in the actual parser help text

zrush-mitre commented on medallion/test/config/test_config_loading.py:
```diff
@@ -0,0 +1,353 @@
+    caplog.set_level(logging.DEBUG, logger="medallion.config")
+    with subtests.test(msg="Used as config file"):
+        assert m_cfg.load_config(conf_file=conf_file) == confdata
+        assert f"load configuration from {str(conf_file)!r}" in caplog.text
```
This line causes the tests to fail on my Windows machine. The conf_file path has 4 slashes in between directories, while caplog is printing with 2 slashes. I'd imagine doing an extra `re.escape` might work, or whatever the opposite is.

maybe-sybr replied:
I think this is actually to do with the repr which is being used to format the `conf_file`. That's only being done to add single quotes so I can just include them in the format string myself. I just sanity checked with this, so I am pretty sure my updated diff will play nice on your Windows box :)

```python
>>> backslash = chr(92)
>>> print(backslash)
\
>>> print(repr(backslash))
'\\'
>>> print(f"{backslash!r}")
'\\'
>>> print(f"--{backslash!r}--")
--'\\'--
```

zrush-mitre commented on medallion/test/config/test_config_loading.py:
```diff
@@ -0,0 +1,353 @@
+    caplog.clear()
+    with subtests.test(msg="Used from config dir"):
+        assert m_cfg.load_config(conf_dir=tmp_path) == confdata
+        assert f"and {str(tmp_path)!r}" in caplog.text
```
Similar to above, the left side is printing with more slashes than the right on Windows (aren't paths between different OSes wonderful?)

maybe-sybr replied:
As above, please resolve if it's fixed!

zrush-mitre commented on medallion/test/test_startup.py:
```diff
@@ -0,0 +1,338 @@
+    parser = medallion.scripts.run._get_argparser()
+    with subtests.test(msg="--conf-dir omitted"):
+        args = parser.parse_args([])
+        assert args.conf_dir == medallion.config.DEFAULT_CONFDIR
```
This test seems to fail when environment variables are set on Windows. `args.conf_dir` comes out as the environment variable I set for `MEDALLION_CONFDIR`.

maybe-sybr replied:
Urgh, I made a bunch of patches for `os.environ` but must have forgotten to do that for these ones. TBH I think a module level fixture for the two new test modules would be more sensible so that we don't need to think about it in future changes.

maybe-sybr replied:
Okay, I believe this (and the variants commented below) should be fixed now

zrush-mitre commented on medallion/test/test_startup.py:
```diff
@@ -0,0 +1,338 @@
+                "sys.argv", ["ARGV0", config_file_arg]
+            ):
+                medallion.scripts.run.main()
+            mock_load_config.assert_called_once_with(
```
Fails with Windows system variables set. It expects `medallion.config.DEFAULT_CONFDIR` on the second call, but receives my environment variable instead.

zrush-mitre commented on medallion/test/test_startup.py:
```diff
@@ -0,0 +1,338 @@
+                "sys.argv", ["ARGV0", "--conf-file", config_file_arg]
+            ):
+                medallion.scripts.run.main()
+            mock_load_config.assert_called_once_with(
```
Same as above

maybe-sybr commented:
@zrush-mitre - I've fixed up those issues (will need confirmation on the windows path rendering) and rebased on top of master. LMK if there's anything I missed :)

zrush-mitre commented:
@maybe-sybr Looking good! LGTM!

maybe-sybr commented:
Thanks @zrush-mitre :)
